dg.go

```diff
@@ -1,315 +1,464 @@
-        }
-    }
-    for toNodeID := range d.dg.connectionsFromNode[d.id] {
-        delete(d.dg.connectionsToNode[toNodeID], d.id)
-    }
-    delete(d.dg.connectionsFromNode, d.id)
-    for fromNodeID := range d.dg.connectionsToNode[d.id] {
-        delete(d.dg.connectionsFromNode[fromNodeID], d.id)
-    }
-    delete(d.dg.connectionsToNode, d.id)
-    delete(d.dg.nodes, d.id)
-    d.deleted = true
-    return nil
+// dependencyResolved is used to notify a node that one of its dependencies have had their resolution
+// status set. Once all dependencies are resolved, the node is set as finalized (ready for processing).
+// Caller should have appropriate mutex locked before calling.
+func (n *node[NodeType]) dependencyResolved(dependencyNodeID string, dependencyResolution ResolutionStatus) error {
+	if n.deleted {
+		return &ErrNodeDeleted{n.id}
+	}
+	if dependencyResolution == Waiting {
+		// Illegal state
+		return ErrNotifiedOfWaiting{n.id, dependencyNodeID}
+	}
+	dependencyType, isOutstandingDependency := n.outstandingDependencies[dependencyNodeID]
+	if !isOutstandingDependency {
+		// Now determine if the missing item was because the dependency was already resolved, or
+		// because there was never a connection.
+		_, isConnected := n.dg.connectionsToNode[n.id][dependencyNodeID]
+		if isConnected {
+			// As designed, this is an internal function. So we guard against this in resolveNode.
+			panic(ErrDuplicateDependencyResolution{n.id, dependencyNodeID})
+		} else {
+			panic(ErrConnectionDoesNotExist{dependencyNodeID, n.id})
+		}
+	}
+	delete(n.outstandingDependencies, dependencyNodeID)
+	if dependencyType == ObviatedDependency {
+		return nil // Nothing to do.
+	}
+	// If resolution is unresolvable, fail if current type is AND, or if there are no remaining OR dependencies.
+	// Treat as resolved if is just a completion dependency.
+	if dependencyResolution == Unresolvable && dependencyType != CompletionDependency {
+		// Check for the failure case.
+		if dependencyType == AndDependency || !n.hasOutstandingDependency(OrDependency) {
+			// Missing requirement. Mark as unresolvable, which propagates to outbound connections.
+			n.ready = true
+			n.dg.readyForProcessing[n.id] = n
+			return n.resolveNode(Unresolvable)
+		}
+	} else {
+		var hasOrDependency bool
+		if dependencyType == OrDependency {
+			// No matter what, mark other ORs as obviated. If there are no AND dependencies left, mark as ready.
+			n.markOrsObviated()
+			hasOrDependency = false // This resolved all outstanding ORs.
+		} else {
+			hasOrDependency = n.hasOutstandingDependency(OrDependency)
+		}
+		hasAndDependency := n.hasOutstandingDependency(AndDependency) || n.hasOutstandingDependency(CompletionDependency)
+		// Now determine if it's ready to be finalized (no more deferred dependencies).
+		if !(hasAndDependency || hasOrDependency) {
+			// Mark as ready for processing internally and in the DAG.
+			n.ready = true
+			n.dg.readyForProcessing[n.id] = n
```

Comment:
Do we need both of these?  `n.ready` strikes me as likely to be redundant (because `n` will come from `readyForProcessing`, right?)  (Moreover, getting rid of `n.ready` will address my earlier concern about the getter function for it....)